jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java

```diff
@@ -379,9 +377,8 @@ public Path getPath()
     @Override
     public InputStream getInputStream() throws IOException
     {
-        // Use a FileInputStream rather than Files.newInputStream(path)
-        // since it produces a stream with a fast skip implementation
-        return new FileInputStream(getFile());
+        // TODO: investigate if SPARSE use for default FileSystem usages is worth it
+        return Files.newInputStream(path, StandardOpenOption.READ);
```

Comment:
I think because we know this is giving us an poor stream we should do:
```java
File file = getFile();
if (file != null)
    return new FileInputStream(file);
return Files.newInputStream(path, StandardOpenOption.READ);
```

I still think that we should probably only put that fix in for 9.4.20 to give us time to more thoroughly review/test the range changes

Reply:
That `getFile()` breaks things, the very things this PR fixes for, we can't speculatively ask the `Path.getFile()`.
We can't use it, even for 9.4.20, as that is not a valid fix for `PathResource`.
It will also never return null, it will throw _various_ exceptions (depending on the `java.nio.files.FileSystem` implementation that the `Path` belongs to).

Also, with the changes in this PR, that `PathResource.getInputStream()` is the last possible choice (in logic).

In the lookup, you have 3 different `ByteBuffer` attempts (direct, mapped, indirect), a `ReadableByteChannel`, `SeekableByteChannel`, another `ByteBuffer` (from cached `HttpContent`), and then finally, at the end, a `InputStream`.

In the case of real files on a real OS file system, using the Default `java.nio.files.FileSystem` implementation, on all OSs, on all JVMs, that final `InputStream` option is never hit.  (this is important to recognize)

The `InputStream` option is only hit on `Path` objects that are _not_ on any of the Default `java.nio.files.FileSystems`, that exceed caching max sizes, and have no `SeekableByteChannel` option (which is a contract violation of `Files.newByteChannel(Path, OpenOption[])` btw).

Also, finally, the slow / poor stream thing you are working to fix is `java.nio.files.FileSystem` dependent, only the Default `java.nio.files.FileSystem` implementation returns that known slow implementation.   Using `Files.newInputStream(Path, OpenOption[])` on a non default `java.nio.files.FileSystem` returns something completely different.

Reply:
If the `Path` in the `PathResource` doesn't belong to the `FileSytem.getDefault()` then according to the spec (and javadoc) it is not interoperable with `java.io.File`.

If you insist on `getFile()` then we have to guard it with checks and validation on the `Path` object and it's associated `FileSystem`.

We have to get the `Path.getFileSystem()` and compare it somehow against the `FileSystems.getDefault()` (I've never done this, so I'm not sure how successful this will be with say a straight `Path.getFileSystem().equals(FileSystem.getDefault())` style test).    

But if the user (or JVM environment) has declared the `java.nio.file.spi.DefaultFileSystemProvider` system property, will that compare still work? (do we care?) - custom default FileSystems are often done on Mac OS and NTFS to let Java gain access to the extra metadata that those file systems provide.

Will this test work on Android with it's custom `java.nio.files.FileSystem` implementation? (do we care?)

Reply:
Hmmm... ok.